Improve the performance of writing attributes

eugeneius opened the pull request:
Using a similar approach to 08576b94ad4f19dfc368619d7751e211d23dcad8, this change adds a new internal `_write_attribute` method which bypasses the code that checks for attribute aliases and custom primary keys.

We can use this method instead of `write_attribute` when we know that we have the name of the actual column to be updated and not an alias.

This makes writing an attribute with `attribute=` about 18% faster.

## Benchmark:

```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", github: "rails/rails"
  gem "arel", github: "rails/arel"
  gem "sqlite3"
  gem "benchmark-ips"
end

require "active_record"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end
end

class Post < ActiveRecord::Base
end

post = Post.new(id: 1)

Benchmark.ips do |x|
  x.report("attribute=") { post.id = post.id + 1 }
end
```

### Before:

    Warming up --------------------------------------
              attribute=    25.889k i/100ms
    Calculating -------------------------------------
              attribute=    290.946k (± 3.1%) i/s -      1.476M in   5.077036s

### After:

    Warming up --------------------------------------
              attribute=    30.056k i/100ms
    Calculating -------------------------------------
              attribute=    345.088k (± 4.8%) i/s -      1.743M in   5.064264s

eugeneius committed:
Improve the performance of writing attributes

Using a similar approach to 08576b94ad4f19dfc368619d7751e211d23dcad8,
this change adds a new internal `_write_attribute` method which bypasses
the code that checks for attribute aliases and custom primary keys.

We can use this method instead of `write_attribute` when we know that we
have the name of the actual column to be updated and not an alias.

This makes writing an attribute with `attribute=` about 18% faster.

Benchmark:

```
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", github: "rails/rails"
  gem "arel", github: "rails/arel"
  gem "sqlite3"
  gem "benchmark-ips"
end

require "active_record"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end
end

class Post < ActiveRecord::Base
end

post = Post.new(id: 1)

Benchmark.ips do |x|
  x.report("attribute=") { post.id = post.id + 1 }
end
```

Before:

    Warming up --------------------------------------
              attribute=    25.889k i/100ms
    Calculating -------------------------------------
              attribute=    290.946k (± 3.1%) i/s -      1.476M in   5.077036s

After:

    Warming up --------------------------------------
              attribute=    30.056k i/100ms
    Calculating -------------------------------------
              attribute=    345.088k (± 4.8%) i/s -      1.743M in   5.064264s

matthewd commented on activerecord/lib/active_record/attribute_methods/write.rb:
```diff
@@ -17,7 +17,7 @@ def define_method_attribute=(name)
             generated_attribute_methods.module_eval <<-STR, __FILE__, __LINE__ + 1
               def __temp__#{safe_name}=(value)
                 name = ::ActiveRecord::AttributeMethods::AttrNames::ATTR_#{safe_name}
```
As we're in the vicinity and squeezing performance on this method... I wonder how this constant lookup compares to `#{name.inspect}.freeze`

eugeneius replied:
They seem to be equivalent, or at least close enough:

### Before:

```
Warming up --------------------------------------
           attribute   134.983k i/100ms
          attribute=    35.139k i/100ms
Calculating -------------------------------------
           attribute      2.338M (± 3.3%) i/s -     11.744M in   5.029545s
          attribute=    401.747k (± 3.4%) i/s -      2.038M in   5.079005s
```

### After:

```
Warming up --------------------------------------
           attribute   132.856k i/100ms
          attribute=    34.014k i/100ms
Calculating -------------------------------------
           attribute      2.337M (± 3.3%) i/s -     11.691M in   5.008654s
          attribute=    399.083k (± 2.8%) i/s -      2.007M in   5.032575s
```

Without `freeze`, for comparison:

```
Warming up --------------------------------------
           attribute   133.407k i/100ms
          attribute=    33.575k i/100ms
Calculating -------------------------------------
           attribute      2.201M (± 2.9%) i/s -     11.073M in   5.034287s
          attribute=    388.022k (± 2.5%) i/s -      1.947M in   5.021768s
```

It _might_ be worth switching to the form you suggested, as it's much simpler to read and avoids creating a bunch of extra constants. I'd see this as a different patch, but if you feel strongly about it I'll wedge it in here 😄

matthewd commented on activerecord/lib/active_record/attribute_methods/write.rb:
```diff
@@ -36,8 +36,7 @@ def write_attribute(attr_name, value)
         end
 
         name = self.class.primary_key if name == "id".freeze && self.class.primary_key
```
Looking at this here, I also wonder what the implications would be of making `id` an "ordinary" attribute_alias.(But that's not related to this PR.)

eugeneius replied:
You read my mind (or my `git stash`)!

It won't be possible until I revive https://github.com/rails/rails/pull/28395 so that `sync_with_transaction_state` is called when accessing a custom primary key. Doing that is blocked on this PR, because I want to use `_write_attribute` in the implementation.

Another thing I want to try is moving the attribute alias checks to a separate module that's only included in the model when the first alias is added - essentially a deopt for the unlikely branch, as most models (in my experience) don't have any aliases.

matthewd commented on activerecord/lib/active_record/attribute_methods/write.rb:
```diff
@@ -36,8 +36,7 @@ def write_attribute(attr_name, value)
+        _write_attribute(name, value)
       end
 
       def raw_write_attribute(attr_name, value) # :nodoc:
```
Seems slightly unfortunate (read: confusing) to have both `raw_write_attribute` and `_write_attribute`, but I don't have any concrete suggestions for avoiding it.

eugeneius replied:
Before https://github.com/rails/rails/pull/29464 the two methods were [branches inside `write_attribute_with_type_cast`](https://github.com/rails/rails/blob/v5.1.1/activerecord/lib/active_record/attribute_methods/write.rb#L55-L59), which I liked even less, but I agree it's not great.

Maybe instead of having a separate method, we could decorate the value with a "don't typecast me" wrapper, and push the conditional into the AttributeSet class?

eugeneius replied:
Renaming `raw_write_attribute` to `write_attribute_without_type_cast` is probably the simplest thing we could do here to make the interface less ambiguous.

matthewd reviewed (commented):
How sure are we that the inheritance and timestamp columns aren't aliases? Come to think of it, how sure are we about `primary_key`?

eugeneius commented:
> How sure are we that the inheritance and timestamp columns aren't aliases?

The inheritance-related code changing here is gated by `finder_needs_type_condition?`, which boils down to ["is `inheritance_column` one of the columns on the table?"](https://github.com/rails/rails/blob/v5.1.1/activerecord/lib/active_record/inheritance.rb#L78) If `inheritance_column` is an alias, the code won't run.

The timestamp-related code similarly [operates directly](https://github.com/rails/rails/blob/v5.1.1/activerecord/lib/active_record/timestamp.rb#L58)  on [column names](https://github.com/rails/rails/blob/v5.1.1/activerecord/lib/active_record/timestamp.rb#L62), so even if `created_at` or `updated_at` are aliased to some other column, it won't be picked up by the code modified here.

> Come to think of it, how sure are we about `primary_key`?

I hadn't thought about this case, but at minimum you'll get inconsistent behaviour because all of the other primary key methods are not alias-aware, e.g. `id` calls through to `_read_attribute`: https://github.com/rails/rails/blob/v5.1.1/activerecord/lib/active_record/attribute_methods/primary_key.rb#L20

From a quick search, these two places seem to rely on `primary_key` being a real column:
https://github.com/rails/rails/blob/v5.1.1/activerecord/lib/active_record/attribute_methods/primary_key.rb#L85
https://github.com/rails/rails/blob/v5.1.1/activerecord/lib/active_record/model_schema.rb#L396

eugeneius committed:
Rename raw_write_attribute to write_attribute_without_type_cast

This name more accurately describes what the method does, and also
disambiguates it from `_write_attribute`, which ignores aliases.

We can also make the method private, since it's not public API and only
called from one place - `update_columns` - without an explicit receiver.

eugeneius commented:
I've renamed `raw_write_attribute` to `write_attribute_without_type_cast` here to disambiguate it from the new `_write_attribute` method.

@matthewd do you have any more feedback here?

matthewd commented:
Nope, looks great thanks!

rafaelfranca commented:
I just noticed this optimization broke an API we had before that overriding `write_attribute` would also change the behavior of `attribute=`. Now people need to override `_write_attribute` that is private API to get the same behavior, or both methods, wish seems not ideal. Can we keep the opmization but keep the same API?
